src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,434 @@
+        private const int AutosyncIntervalMinutes = 10;
+
+        /// <summary>Amount of headers that should be cached until we stop syncing from the peer.</summary>
+        private const int CacheSyncHeadersThreshold = 2000;
+
+        /// <summary>List of block headers that were not yet consumed by <see cref="ConsensusManager"/>.</summary>
+        /// <remarks>Should be protected by <see cref="asyncLock"/>.</remarks>
+        private readonly List<BlockHeader> cachedHeaders;
+
+        /// <summary>Protects access to <see cref="cachedHeaders"/>.</summary>
+        private readonly AsyncLock asyncLock;
+
+        public ConsensusManagerBehavior(ConcurrentChain chain, IInitialBlockDownloadState initialBlockDownloadState, ConsensusManager consensusManager, IPeerBanning peerBanning, IConnectionManager connectionManager, ILoggerFactory loggerFactory)
+        {
+            this.loggerFactory = loggerFactory;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.chain = chain;
+            this.connectionManager = connectionManager;
+            this.peerBanning = peerBanning;
+
+            this.cachedHeaders = new List<BlockHeader>();
+            this.asyncLock = new AsyncLock();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        ///  <inheritdoc />
+        protected override void AttachCore()
+        {
+            this.logger.LogTrace("()");
+
+            // Initialize auto sync timer.
+            this.autosyncTimer = new Timer(async (o) =>
+            {
+                this.logger.LogTrace("()");
+
+                await this.TrySyncAsync(true).ConfigureAwait(false);
+
+                this.logger.LogTrace("(-)");
+            }, null, 0, (int)TimeSpan.FromMinutes(AutosyncIntervalMinutes).TotalMilliseconds);
+
+            if (this.AttachedPeer.State == NetworkPeerState.Connected)
+                this.AttachedPeer.MyVersion.StartHeight = this.consensusManager.Tip.Height;
+
+            this.AttachedPeer.StateChanged.Register(this.OnStateChangedAsync);
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync, true);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Presents cached headers to <see cref="ConsensusManager"/> from the cache if any and removes consumed from the cache.</summary>
+        /// <param name="newTip">New consensus tip.</param>
+        public async Task<ConnectNewHeadersResult> ConsensusTipChangedAsync(ChainedHeader newTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(newTip), newTip);
+
+            ConnectNewHeadersResult result = null;
+            bool syncRequired = false;
+
+            using (await this.asyncLock.LockAsync().ConfigureAwait(false))
+            {
+                if (this.cachedHeaders.Count != 0)
```

Comment:
There are too many nested ifs here. I would return straight here by checking this.cachedHeaders.Any():

```
if (!this.cachedHeaders.Any())
return;
```

Reply:
Also, why bother opening a `using` statement when the cached header count is 0? `syncRequired` will be false anyway.

Reply:
>Also, why bother opening a using statement when the cached header count is 0? 

Because `cachedHeaders` should be accessed under the lock